keps/sig-node/runtime-class-scheduling.md

```diff
@@ -0,0 +1,347 @@
+    // NodeSelectorTerm.
+    // +optional
+    NodeSelectorTerm []corev1.NodeSelectorRequirement
+
+    // tolerations adds tolerations to pods running with this RuntimeClass.
+    // +optional
+    Tolerations []corev1.Toleration
+}
+```
+
+**NodeSelector vs. NodeAffinity vs. NodeSelectorRequirement**
+
+The PodSpec's `NodeSelector` is a label `map[string]string` that must exactly
+match a subset of node labels. [NodeAffinity][] is a much more complex and
+expressive set of requirements and preferences. NodeSelectorRequirements are a
+small subset of the NodeAffinity rules, that place intersecting requirements on
+a NodeSelectorTerm. There is no way to logically intersect NodeAffinities, and
+intersecting the NodeAffinity.NodeSelector multiplies the number of terms. Using
+[native scheduler support](#runtimeclass-aware-scheduling) greatly simplifies
+the intersection logic.
+
+NodeSelectorRequirements fall in the sweet spot of being expressive enough to
+represent a variety of topologies while still being logically intersectable with
+other pod scheduling constraints. Including more advanced scheduling preferences
+for a RuntimeClass is a use case that is better suited to
+[SchedulingPolicy](#scheduling-policy).
+
+[NodeAffinity]: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity
+
+**Tolerations**
+
+While NodeSelectors and labels are used for steering pods _towards_ nodes,
+[taints and tolerations][taint-and-toleration] are used for steering pods _away_
+from nodes. If every pod had a RuntimeClass and every RuntimeClass had a strict
+NodeSelector, then RuntimeClasses could use non-overlapping selectors in place
+of taints & tolerations. However the same could be said of regular pod
+selectors, yet taints & tolerations are still a useful addition. Examples of
+[use cases](#user-stories) for including tolerations in RuntimeClass topology
+inculde:
+
+- Tainting Windows nodes with `windows.microsoft.com` to keep default linux pods
+  away from the nodes. Windows RuntimeClasses would then need a corresponding
+  toleration.
+- Tainting "sandbox" nodes with `sandboxed.kubernetes.io` to keep services
+  providing privileged node features away from sandboxed workloads. Sandboxed
+  RuntimeClasses would need a toleration to enable them to be run on those
+  nodes.
+
+### RuntimeController
+
+RuntimeController is a new in-tree admission plugin that should eventually be
+enabled on almost every Kubernetes clusters. The role of the controller for
+scheduling is to merge the topology constraints from the RuntimeClass into the
+PodSpec. Eventually, the controller's responsibilities may grow, such as to
+merge in [pod overhead][] or validate feature compatibility.
+
+Note that the RuntimeController is not needed if we implement [native scheduler
+support](#runtimeclass-aware-scheduling).
+
+#### Mix-in
+
+The RuntimeClass topology is merged with the pod's NodeSelector & Tolerations.
```

Comment:
Suppose a pod specifies RuntimeClass A, and A's nodeSelectorReqs and tolerations gets mixed in.

- If A's spec got changed, I believe admission plugin won't step in, say the pod's spec will be out of sync, right?
- If A's spec got changed and the pod gets updated (still refer to RuntimeClass A) afterwards, it seems we have to take out A's old nodeSelectorReqs and tolerations first, and then merge in A's new nodeSelectorReqs and tolerations?